spaceship/lib/spaceship/two_step_or_factor_client.rb

```diff
@@ -172,23 +183,68 @@ def handle_two_factor(response, depth = 0)
       return true
     end
 
-    def get_id_for_number(phone_numbers, result)
+    # extracted into its own method for testing
+    def ask_for_2fa_code(text)
+      ask(text)
+    end
+
+    def phone_id_from_number(phone_numbers, phone_number)
+      characters_to_remove_from_phone_numbers = ' \-()"'
+
+      # start with e.g. +49 162 1234585 or +1-123-456-7866
+      phone_number = phone_number.tr(characters_to_remove_from_phone_numbers, '')
+      # cleaned: +491621234585 or +11234567866
+
       phone_numbers.each do |phone|
-        phone_id = phone['id']
-        return phone_id if phone['numberWithDialCode'] == result
+        # rubocop:disable Style/AsciiComments
+        # start with: +49 •••• •••••85 or +1 (•••) •••-••66
+        number_with_dialcode_masked = phone['numberWithDialCode'].tr(characters_to_remove_from_phone_numbers, '')
+        # cleaned: +49•••••••••85 or +1••••••••66
+        # rubocop:enable Style/AsciiComments
+
+        maskings_count = number_with_dialcode_masked.count('•') # => 9 or 8
+        pattern = /^([0-9+]{2,4})([•]{#{maskings_count}})([0-9]{2})$/
+        replacement = "\\1([0-9]{#{maskings_count - 1},#{maskings_count}})\\3"
+        number_with_dialcode_regex_part = number_with_dialcode_masked.gsub(pattern, replacement)
+        # => +49([0-9]{8,9})85 or +1([0-9]{7,8})66
+
+        backslash = '\\'
+        number_with_dialcode_regex_part = backslash + number_with_dialcode_regex_part
+        number_with_dialcode_regex = /^#{number_with_dialcode_regex_part}$/
+        # => /^\+49([0-9]{8})85$/ or /^\+1([0-9]{7,8})66$/
+
+        return phone['id'] if phone_number =~ number_with_dialcode_regex
+        # +491621234585 matches /^\+49([0-9]{8})85$/
       end
+
+      # Handle case of phone_number not existing in phone_numbers because ENV var is wrong or matcher is broken
+      raise Tunes::Error.new, %(
+Could not find a matching phone number to #{phone_number} in #{phone_numbers}.
```

Comment:
We may want to mask the `phone_number` here 🤔 The `phone_numbers` also prints out as a Ruby object. Do we want to make that a little more CLI friendly? 😇
```sh
Could not find a matching phone number to <plain text phone number> in [{"numberWithDialCode"=>"+1 (•••) •••-••98", "pushMode"=>"sms", "obfuscatedNumber"=>"(•••) •••-••98", "id"=>1}].
```

Reply:
Problem is that we actually need the phone number here to be able to debug the problem. We could add something like "If you are uncomfortable sharing your phone number, please switch a few digits or replace them with `0`."

Reply:
@janpio Could we only not mask if verbose is on? 🤔

Reply:
We could, but then we would have to add some information that the users needs to run `--verbose` before reporting the bug.

Reply:
I prefer not masking here and leave the masking to users as we do with team ID and name etc.